### ISSUE TITLE
Collapsing translations sometimes loses sight of the sentence block

### ISSUE THREAD

**Issue by trang** [Tatoeba/tatoeba2#2093]
Steps to reproduce:

- Do some [random search](https://tatoeba.org/eng/sentences/search?query=&from=und&to=und&user=&orphans=no&unapproved=no&has_audio=&tags=&list=&native=&trans_filter=limit&trans_to=und&trans_link=&trans_user=&trans_orphan=&trans_unapproved=&trans_has_audio=&sort=random&sort_reverse=) or any search that can find a sentence with many translations.
- Find a sentence in the middle of the page that has many translations and expand to see all its translations.
- Scroll down till the "Fewer translations" button gets towards the top of the page (or scroll down until you see the button, if there are really a lot of translations).
- Click "Fewer translations" to collapse the translations.
- The sentence block disappears from the visible part of the screen.

Expected:
- When collapsing a long list of translations, the screen should be re-positioned to display again the first translations.

**Comment by blackslate** [comment]
Expanding and collapsing sentences works differently in different browsers.

I cannot reproduce this in Google Chrome version 83.0.4103.61 on the online site. This search ...
https://tatoeba.org/eng/sentences/search?query=good&from=eng&to=und
... provides 10 expressions containing the word "good", seven of which have over 100 translations. On Google Chrome, these expressions are displayed in one long list, with no button that says Show NN More Translations.

On Firefox 77.0.1, the translations are collapsed and the Show NN More Translations button appears, but scrolling until the Fewer Translation button appears and then clicking on it collapses the list correctly, in most cases, so that the button Show NN More Translations for that particular sentence appears under the cursor.

The cases where this does not happen are sentences with only a few More Translations, such as "Good show!" and "Good riddance!"

In Opera 68.0.3618.125, it seems that the only sentence at this link that misbehaves is  "Good riddance", and this only when the Fewer Translations button is located in the lower third of the window, or when more than the last 14 translations are visible.

**Comment by blackslate** [comment]
Step one produces no results in my local version of Tatoeba running in Vagrant:

1. Do some random search or any search that can find a sentence with many translations.

In my local version, I have added a number of sentences, all of which include the word "sentence" but when I use the search bar to look for phrases which contain the word "sentence", I get the result:
  
  No results found for: sentence

Simply visiting the site `http://localhost:8080/` is enough to display a single random sentence, which may have More Translations to show, but in order to reproduce the issue, I need to display several sentences.

**Comment by trang** [comment]
> On Google Chrome, these expressions are displayed in one long list, with no button that says Show NN More Translations.

This is probably because you are authenticated. Authenticated users have a different design for the sentences, which we call the "old design". Also, in the old design, by default there is no button to expand/collapse. This button only appears if you go to [your Settings](https://tatoeba.org/eng/user/settings) and enable the option "Display a link to expand/collapse translations when there are too many translations".

But we want to fix this on the new design only, not the old one (which is planned to be removed completely at some point). So for your tests, you should either remain non-authenticated or you can enable the option "Display sentences with the new design..." in your Settings.

> but scrolling until the Fewer Translation button appears and then clicking on it collapses the list correctly

The issue is indeed not consistently reproducible. It depends on the sentence you're expanding. But to remove the random factor, you can go to this list instead:
https://tatoeba.org/eng/sentences_lists/show/166576/und

Then use the first sentence ("The dog will bark."). With this, I could reproduce the issue consistently, on every browser. Here are some screenshots: https://imgur.com/a/9YdxeEa.

To reproduce locally from scratch, the steps would be:

1. Go to the lists section (http://localhost:8080/eng/sentences_lists/index).
2. From there, create a new list (in the sidebar, there's a box "Create a new list).
3. Add a few sentences in the list.
4. Add 10 translations to the latest sentence you created.
5. Add 2-3 translations to the other sentences.
6. In the sidebar, in the "Action" > "Show translations :" dropdown list, select "All languages". This is just to ensure that if you refresh the page, all translations are displayed.
7. Expand the first sentence in the list (it should be the one that has the most translations).
8. Scroll down till the "Fewer translations" button gets towards the top of the page.
9. Click "Fewer translations" to collapse the translations.

> No results found for: sentence

To be able to find sentences in the search, you will need to run manually the indexation from the VM. The sentences are not indexed automatically in the local installation.

```
$ cd /path/to/imouto/
$ vagrant ssh

// once connected to the VM
$ cd Tatoeba
$ sudo bin/cake sphinx_indexes update main
```

**Comment by blackslate** [comment]
Hi Trang, 

Is there a way to run the application using non-minified code? 

I'm not sure how helpful I will be able to be on this issue. I've
searched the code in /imouto/Tatoeba/webroot/js/ for the strings
"collapse" and "fewer", hoping that this will give me some insight into
what is happening. I see that  in
/imouto/Tatoeba/webroot/js/directives/sentence-and-translations.dir.js,
the variables vm.(in)directTranslations are modified,  and so,
presumably the layout changes automatically. However, as I have said
before, I have little experience in using Angular, so I do not know how
to influence the way it scrolls the page. 

The solution that I imagine, if I knew where to put it, would be to use
getBoundingClientRect twice: first to determine where the Show Fewer
button is before the block of translations is collapsed, and a second
time to find where the Show ... More ... button is after the block is
collapsed. I would then scroll the page so that the button is in the
same position as it was before it was clicked. However, it is possible
that Angular already provides an automatic way of doing this, but that
this has not been successfully implemented. If this is the case, my
solution is likely to be an un-Angular hack : )  

In any case, stepping through minified code is not an efficient way to
find where to make changes. 

If you think that it is worth our time continuing to work on this
together, I'll be happy to get your feedback. 

James 

On 06.06.2020 19:54, Trang wrote:

>> On Google Chrome, these expressions are displayed in one long list, with no button that says Show NN More Translations.
> 
> This is probably because you are authenticated. Authenticated users have a different design for the sentences, which we call the "old design". Also, in the old design, by default there is no button to expand/collapse. This button only appears if you go to your Settings [1] and enable the option "Display a link to expand/collapse translations when there are too many translations". 
> 
> But we want to fix this on the new design only, not the old one (which is planned to be removed completely at some point). So for your tests, you should either remain non-authenticated or you can enable the option "Display sentences with the new design..." in your Settings.
> 
>> but scrolling until the Fewer Translation button appears and then clicking on it collapses the list correctly
> 
> The issue is indeed not consistently reproducible. It depends on the sentence you're expanding. But to remove the random factor, you can go to this list instead:
> https://tatoeba.org/eng/sentences_lists/show/166576/und 
> 
> Then use the first sentence ("The dog will bark."). With this, I could reproduce the issue consistently, on every browser. Here are some screenshots: https://imgur.com/a/9YdxeEa. 
> 
> To reproduce locally from scratch, the steps would be: 
> 
> * Go to the lists section (http://localhost:8080/eng/sentences_lists/index).
> * From there, create a new list (in the sidebar, there's a box "Create a new list).
> * Add a few sentences in the list.
> * Add 10 translations to the latest sentence you created.
> * Add 2-3 translations to the other sentences.
> * In the sidebar, in the "Action" > "Show translations :" dropdown list, select "All languages". This is just to ensure that if you refresh the page, all translations are displayed.
> * Expand the first sentence in the list (it should be the one that has the most translations).
> * Scroll down till the "Fewer translations" button gets towards the top of the page.
> * Click "Fewer translations" to collapse the translations.
> 
>> No results found for: sentence
> 
> To be able to find sentences in the search, you will need to run manually the indexation from the VM. The sentences are not indexed automatically in the local installation.
> 
> $ cd /path/to/imouto/
> $ vagrant ssh
> 
> // once connected to the VM
> $ cd Tatoeba
> $ sudo bin/cake sphinx_indexes update main
> 
> --
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub [2], or unsubscribe [3].

 

Links:
------
[1] https://tatoeba.org/eng/user/settings
[2]
https://github.com/Tatoeba/tatoeba2/issues/2093#issuecomment-640088898
[3]

**Comment by trang** [comment]
> Is there a way to run the application using non-minified code?

By default it should be running in non-minified mode for the Tatoeba javascript. Some javascript files will however be bundled into one single file.

That is the case for `sentence-and-translations.dir.js`. It is bundled with other js files into a file called `sentence-component.js`. You can check the [asset_compress.ini](https://github.com/Tatoeba/tatoeba2/blob/dev/config/asset_compress.ini) to see which js files are being bundled together.

So if you are trying to put breakpoints in the Debugger from your browser, you would have to look in `/cache_js/sentence-component.js`.

But for this issue, there is no need to do touch the javascript.

> The solution that I imagine, if I knew where to put it, would be to use getBoundingClientRect twice

There is a simpler way: just using an HTML anchor. 

This mechanism is used for instance [on the Wall](https://tatoeba.org/eng/wall/index):
- In the sidebar, there is this list of "Latest messages". 
- If you inspect an element in this list, you'll see that it's basically just an anchor link with `href="#message_{id}"`.
- Clicking on an item will scroll the page to the corresponding message.

So in your case, you would need to do something similar for the sentences. That is:
1.  Finding the file where the HTML of this "Fewer translations" button is defined.
2. Setting the `href` attribute on this button (or rather `ng-href` because you'll need to read a variable from Angular to get the id of the sentence. You'll find other usages of `ng-href` in the file).
3. Setting the proper `id` on the sentence component `<div>`.

> If you think that it is worth our time continuing to work on this together, I'll be happy to get your feedback.

Yes, no worries. It is not expected that you figure out right away what to do on the first issue. We can't escape the learning curve :)

Plus it is interesting for us to see what new developers struggle with. The fact that you didn't have pre-existing data to test the bug and the fact that you were confused with the asset compression are things that we don't always take into account, but should, when involving new developers.

**Comment by blackslate** [comment]
On 07.06.2020 02:02, Trang wrote:

> So in your case, you would need to do something similar for the sentences. That is: 
> 
> * Finding the file where the HTML of this "Fewer translations" button is defined.
> * Setting the href attribute on this button (or rather ng-href because you'll need to read a variable from Angular to get the id of the sentence. You'll find other usages of ng-href in the file).
> * Setting the proper id on the sentence component <div>.

Hi Trang, 

I'm sorry, your coding process is completely alien to me. You are
placing JavaScript code which uses the Angular framework and JQuery
inside PHP files. As I have said before, I am used to writing
JQuery-free code using the React framework in a Node.js environment.
Before I can begin to understand what code needs to be changed where,
I'll need to learn a several new coding paradigms, none of which are
currently useful to me for my other projects. 

In addition, I am not convinced that your HTML anchor approach will
solve the problem. If we set the the href of the page then the scroll of
the page will change every time a Fewer Translations button is clicked,
even when a change of scroll is unnecessary. Besides, setting the href
of the page has a side-effect on the action of browser's back button,
which some users might find undesirable. 

In other words, this is not a "good first issue" as far as I am
concerned. I imagine that someone with a different skill set might
resolve it much more efficiently than I could. 

Sincerely, 

James

**Comment by trang** [comment]
Hi James,

If you prefer to switch to another issue, it's no problem. I will send you by email a list of 3-4 issues and you can tell me if one of them inspires you more.

You brought up a fair point regarding the degradation of user experience by using an anchor link. I honestly didn't think much about it but having tried it now, it does feel quite clunky. So even if you won't be implementing the final solution, you did help significantly with your feedback :)

After some research, I think using [`.scrollIntoView()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) with the options `{behavior: 'smooth', block: 'nearest'}` can make the experience acceptable. It is however not supported by IE but [there is a polyfill](https://github.com/iamdustan/smoothscroll). That's perhaps a direction to investigate for the next person who wants to solves this issue.
